File: Final/Code/Project_4/DFS/MultiGraph/MultiGraph.py
from collections import defaultdict

class Multigraph:
    def __init__(self):
        self.graph = defaultdict(list)

    def add_edge(self, u, v):
        self.graph[u].append(v)

    def dfs_util(self, v, visited):
        visited.add(v)
        print(v, end=' ')
        for neighbor in self.graph[v]:
            if neighbor not in visited:
                self.dfs_util(neighbor, visited)

    def dfs(self, start):
        visited = set()
        print("DFS traversal:")
        self.dfs_util(start, visited)

G = Multigraph()
G.add_edge(0, 1)
G.add_edge(0, 1)  # đa cạnh
G.add_edge(1, 2)
G.add_edge(2, 0)
G.add_edge(2, 3)
G.add_edge(3, 3)  # cạnh khuyên

G.dfs(2)

File: Final/Code/Project_4/DFS/MultiGraph/MultiGraph.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
#include <unordered_set>
using namespace std;

class Multigraph {
private:
    unordered_map<int, vector<int>> adj;

    void dfsUtil(int v, unordered_set<int>& visited) {
        visited.insert(v);
        cout << v << " ";
        for (int neighbor : adj[v]) {
            if (visited.find(neighbor) == visited.end()) {
                dfsUtil(neighbor, visited);
            }
        }
    }

public:
    void addEdge(int u, int v) {
        adj[u].push_back(v); 
    }

    void dfs(int start) {
        unordered_set<int> visited;
        cout << "DFS traversal: ";
        dfsUtil(start, visited);
        cout << endl;
    }
};

int main() {
    Multigraph G;
    G.addEdge(0, 1);
    G.addEdge(0, 1); 
    G.addEdge(1, 2);
    G.addEdge(2, 0);
    G.addEdge(2, 3);
    G.addEdge(3, 3); 

    G.dfs(2);

    return 0;
}